### PR TITLE
feat(core): add IPC channel

### PR DESCRIPTION
**Pull request by lucasfernog**
### What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Docs
- [ ] New Binding issue #___
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

### Does this PR introduce a breaking change?

- [ ] Yes, and the changes were approved in issue #___
- [ ] No

### Checklist
- [ ] When resolving issues, they are referenced in the PR's title (e.g `fix: remove a typo, closes #___, #___`)
- [x] A change file is added if any packages will require a version bump due to this PR per [the instructions in the readme](https://github.com/tauri-apps/tauri/blob/dev/.changes/readme.md).
- [ ] I have added a convincing reason for adding this feature, if necessary

### CODE REVIEW

**Review by wusyong — COMMENTED** [review]
Is this mobile only feature?

**Comment by lucasfernog** [comment]
> Is this mobile only feature?

I've also implemented it for desktop (it can easily be backported to v1) though the main idea here is to use it to send events from the Kotlin and Swift sides to the webview (eventually we'll also need a communication channel between mobile and Rust, but that's a little complicated since it involves static variables).

**Comment by JonasKruckenberg** [comment]
Is this part of the public API?

**Comment by lucasfernog** [comment]
Yeah it is part of the public API @JonasKruckenberg

**Review by chippers — CHANGES_REQUESTED** [review]
I'm not fully up to date with mobile, but I see what you are trying to do with the channel communication here.

Only suggestion would be to cut the channel prefix once, instead of splitting on the string and grabbing the second.

**Review by chippers — DISMISSED** [review]

**Review by chippers — DISMISSED** [review]

**Comment by JonasKruckenberg** [comment]
Can we hold off on merging this until I had the chance to review it tomorrow?

**Comment by lucasfernog** [comment]
Sure @JonasKruckenberg

File: core/tauri/src/api/ipc.rs [hunk 1]
```diff
@@ -10,6 +10,66 @@ use serde::{Deserialize, Serialize};
...
+    let window = command.message.window();
+    let value: String =
+      Deserialize::deserialize(command).map_err(|e| crate::Error::InvalidArgs(name, arg, e))?;
+    if value.starts_with(CHANNEL_PREFIX) {
+      let callback_id: Option<usize> = value
+        .chars()
+        .skip(CHANNEL_PREFIX.len())
+        .collect::<String>()
+        .parse()
+        .ok();
```
Comment (JonasKruckenberg) [inline core/tauri/src/api/ipc.rs]: Maybe use `value.split_once(CHANNEL_PREFIX)`

File: tooling/api/src/tauri.ts [hunk 2]
```diff
@@ -55,6 +55,29 @@ function transformCallback(
   return identifier
 }
 
+class Channel {
```
Comment (JonasKruckenberg) [inline tooling/api/src/tauri.ts]: I would refactor this to be more similar to the `Websocket` or `BroadcastChannel` api i.e. the constructor takes no arguments and you can listen to events by overriding the `on message` property or calling `.addEventListener('message', () => {})`

**Review by JonasKruckenberg — CHANGES_REQUESTED** [review]

File: core/tauri/src/api/ipc.rs (hunk 1 above)
Comment (lucasfernog) [reply core/tauri/src/api/ipc.rs]: Pushed.

File: tooling/api/src/tauri.ts (hunk 2 above)
Comment (lucasfernog) [reply tooling/api/src/tauri.ts]: Good idea, pushed.

**Review by wusyong — APPROVED** [review]

**Review by JonasKruckenberg — APPROVED** [review]

**Pull request merged**